src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,208 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    internal enum DnsMonitorState
+    {
+        Created,
+        Running,
+        Failed,
+        Stopped
+    }
+
+    internal class DnsMonitor
+    {
+        #region static
+        private static string EnsureLookupDomainNameIsValid(string lookupDomainName)
+        {
+            Ensure.IsNotNull(lookupDomainName, nameof(lookupDomainName));
+            Ensure.That(lookupDomainName.Count(c => c == '.') >= 2, "LookupDomainName must have at least three components.", nameof(lookupDomainName));
+            return lookupDomainName;
+        }
+
+        private static string GetParentDomainName(string domainName)
+        {
+            var index = domainName.IndexOf('.');
+            return domainName.Substring(index + 1);
+        }
+        #endregion
+
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly IDnsMonitoringCluster _cluster;
+        private readonly IDnsResolver _dnsResolver;
+        private readonly string _lookupDomainName;
+        private readonly string _parentDomainName;
+        private bool _processDnsResultHasEverBeenCalled;
+        private readonly string _service;
+        private DnsMonitorState _state;
+        private Exception _unhandledException;
+
+        private readonly Action<SdamInformationEvent> _sdamInformationEventHandler;
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, IDnsResolver dnsResolver, string lookupDomainName, IEventSubscriber eventSubscriber, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _dnsResolver = Ensure.IsNotNull(dnsResolver, nameof(dnsResolver));
+            _lookupDomainName = EnsureLookupDomainNameIsValid(lookupDomainName);
+            _cancellationToken = cancellationToken;
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _service = "_mongodb._tcp." + _lookupDomainName;
+            _state = DnsMonitorState.Created;
+
+            eventSubscriber?.TryGetEventHandler(out _sdamInformationEventHandler);
+        }
+
+        // public properties
+        public DnsMonitorState State => _state;
+
+        public Exception UnhandledException => _unhandledException;
+
+        // public methods
+        public void Start()
+        {
+            _state = DnsMonitorState.Running;
+
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception exception)
```

Comment:
Why not just catch everything in Monitor method and keep looping?

Reply:
These are exceptions that should NEVER happen.

Exceptions that are expected are handled in Monitor without exiting the loop.